### PR TITLE
HBE-326 feature: server configuration through GraphQL API

### PR DESCRIPTION
**Pull request by mirarifhasan**
Closes HBE-326

### Description

This PR introduces the ability (in backend service) to update server configurations (`GOOGLE_CLIENT_ID`, `MAILER_SMTP_URL`, `VITE_ALLOWED_AUTH_PROVIDERS`, etc) from the UI (Admin Dashboard).

Graphql Mutation and Query lists: https://linear.app/hopp/issue/HBE-326/shc-hands-on-implementation-server-configuration

Details of backend implementation: https://www.notion.so/hoppscotch/Env-Variable-Configuration-from-UI-609016e231fc4f8ab85d3b2d83926ec1

### Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

### Testing Instruction

1. You can test/run the above query and mutation. 
2. Check the InfraCondig table (in DB) which should reflect the new value.
3. Check after any mutation, the backend server will restart

### Additional Information

Nil

### CODE REVIEW

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        value: process.env.MICROSOFT_CLIENT_ID,
+      },
+      {
+        name: InfraConfigEnum.MICROSOFT_CLIENT_SECRET,
+        value: process.env.MICROSOFT_CLIENT_SECRET,
+      },
+      {
+        name: InfraConfigEnum.SAML_ISSUER,
+        value: process.env.SAML_ISSUER,
+      },
```
Comment (balub) [inline packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:67]: Remove SAML related details from this PR.

**Review by balub — CHANGES_REQUESTED** [review]

**Comment by balub** [comment]
@mirarifhasan can you convert this PR to a draft one.

**Review by balub — CHANGES_REQUESTED** [review]

File: packages/hoppscotch-backend/src/admin/infra.resolver.ts [hunk 2]
```diff
@@ -222,4 +239,76 @@ export class InfraResolver {
...
+    return infraConfigs.right;
+  }
+
+  @Query(() => [String], {
+    description: 'Allowed Auth Provider list',
+  })
+  @UseGuards(GqlAuthGuard, GqlAdminGuard)
+  allowedAuthProviders() {
+    return this.infraConfigService.getAllowedAuthProviders();
+  }
```
[the remark this answers is not in the document]
Comment (mirarifhasan) [reply packages/hoppscotch-backend/src/admin/infra.resolver.ts:266]: @balub In Admin Dashboard, frontend calls this query `allowedAuthProviders`

https://hoppscotchworkspace.slack.com/archives/C04C148N2JK/p1702014906806529?thread_ts=1701954362.399499&cid=C04C148N2JK

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+      await this.prisma.$transaction(async (tx) => {
+        const deleteCount = await tx.infraConfig.deleteMany({
+          where: { name: { in: infraConfigs.map((p) => p.name) } },
+        });
+
+        const createCount = await tx.infraConfig.createMany({
+          data: infraConfigs.map((p) => ({
+            name: p.name,
+            value: p.value,
+          })),
```
[the remark this answers is not in the document]
Comment (mirarifhasan) [reply packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:177]: You're right @balub.
I'll change it

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+    try {
+      const infraConfig = await this.prisma.infraConfig.update({
+        where: { name },
+        data: { value },
+      });
+
+      stopApp();
+
+      return E.right(this.cast(infraConfig));
```
[the remark this answers is not in the document]
Comment (mirarifhasan) [reply packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:152]: I think we can leave it. `update()` method is a helper function for other methods like `enableAndDisableSSO()`.
IMO, good to return from this helper function.

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+      return E.left(AUTH_PROVIDER_NOT_SPECIFIED);
+    }
+
+    const isUpdated = await this.update(
+      InfraConfigEnum.VITE_ALLOWED_AUTH_PROVIDERS,
+      newEnabledAuthProviders.join(','),
+    );
+    if (E.isLeft(isUpdated)) return E.left(isUpdated.left);
+
+    return E.right(true);
```
[the remark this answers is not in the document]
Comment (mirarifhasan) [reply packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:228]: changed a bit

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+
+      stopApp();
+
+      return E.right(true);
+    } catch (e) {
+      return E.left(INFRA_CONFIG_RESET_FAILED);
+    }
+  }
+
+  validateEnvValues(
```
Comment (balub) [inline packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:289]: I think we can add a check for the `MAILER_ADDRESS_FROM` variable here too.

**Review by balub — CHANGES_REQUESTED** [review]

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts (hunk 6 above)
Comment (balub) [reply packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:289]: How about `VITE_ALLOWED_AUTH_PROVIDERS` too?

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    for (let i = 0; i < infraConfigs.length; i++) {
+      if (infraConfigs[i].name === InfraConfigEnumForClient.MAILER_SMTP_URL) {
+        const isValidUrl = validateSMTPUrl(infraConfigs[i].value);
+        if (!isValidUrl) return E.left(INFRA_CONFIG_INVALID_INPUT);
+      } else if (
+        infraConfigs[i].name === InfraConfigEnumForClient.MAILER_ADDRESS_FROM
+      ) {
+        const isValidEmail = validateEmail(infraConfigs[i].value);
+        if (!isValidEmail) return E.left(INFRA_CONFIG_INVALID_INPUT);
+      }
```
Comment (balub) [inline packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:304]: We should use a `switch` here.

**Review by balub — CHANGES_REQUESTED** [review]

**Review by balub — APPROVED** [review]

**Review by balub — APPROVED** [review]

**Review by balub — APPROVED** [review]

**Review by ankitsridhar16 — APPROVED** [review]

**Review by ankitsridhar16 — APPROVED** [review]

**Pull request merged**